test/sys/test_socket.rs

```diff
@@ -66,6 +68,37 @@ pub fn test_path_to_sock_addr() {
     assert_eq!(addr.path(), Some(actual));
 }
 
+fn calculate_hash<T: Hash>(t: &T) -> u64 {
+    let mut s = DefaultHasher::new();
+    t.hash(&mut s);
+    s.finish()
+}
+
+#[test]
+pub fn test_addr_equality_path() {
+    let path = "/foo/bar";
+    let actual = Path::new(path);
+    let addr1 = UnixAddr::new(actual).unwrap();
+    let mut addr2 = addr1.clone();
+
+    addr2.0.sun_path[10] = 127;
+
+    assert_eq!(addr1, addr2);
+    assert_eq!(calculate_hash(&addr1), calculate_hash(&addr2));
+}
+
+#[cfg(target_os = "linux")]
+#[test]
+pub fn test_addr_equality_abstract() {
+    let name = String::from("nix\0abstract\0test");
+    let addr1 = UnixAddr::new_abstract(name.as_bytes()).unwrap();
+    let mut addr2 = addr1.clone();
+    addr2.0.sun_path[18] = 127;
```

Comment:
Why should these two addrs be equal?  I thought that in an abstract socket, all bytes were relevant?

Reply:
It actually looks like our test code was wrong here, as it truncated abstract address names to the end of whatever string was passed in, when really it should be the entire array (minus the first byte). So I've also modified that behavior here. Need to add a CHANGELOG entry if you agree this new implementation is the correct one.